refactor: half types and mathematics assembly references

NoelStephensUnity opened the pull request:
This PR refactors the more recent NetworkTransform updates (#2388) by replacing some of the half precision implementations with that of the Unity.Mathematics types. It also adds all of the Unity.Mathematics types to the NGO serialization pipeline.


## Testing and Documentation

- Includes integration test and updates to existing tests.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
update

adding asmdef references to Unity.Mathematics

NoelStephensUnity committed:
refactor

updated HalfVector3 and HalfVector4 with Unity.Mathematics types.
removed classes no longer needed.
renamed HalfVector3DeltaPosition to NetworkDeltaPosition and made that use HalfVector3.

NoelStephensUnity committed:
style

removing whitespace

NoelStephensUnity committed:
update

setting the Vector4 to size of 4 (not 3).
Updating the teleport fix for testing purposes.

NoelStephensUnity committed:
test

mirroring PR-2451 changes to NetworkTransformStateTests for the teleport fix.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/half-types-and-mathematics-references

NoelStephensUnity committed:
update

Added all Mathematics types to the serialization pipeline.

NoelStephensUnity committed:
update

Added arrays for all Unity.Mathematics types.
Fixed some managed read/write issues with Unity.Mathematics types.

NoelStephensUnity committed:
test

Added the tests for the Unity.Mathematics types.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/half-types-and-mathematics-references

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -1185,6 +1186,272 @@ public struct ForFixedStrings
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void WriteValue(Ray2D[] value) => WriteUnmanaged(value);
 
+        #region UnityMathematics As Value
```
we do not use any `#region` blocks as standard

NoelStephensUnity replied:
Thanks for pointing that out.  I was aware of this, but am using it currently to separate the additions.
I still have to add all of the XML API documentation to each added method which is why it is still in draft.
_(will make sure to remove them when I am done)_

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -1185,6 +1186,272 @@ public struct ForFixedStrings
+        #endregion
+
+
+        #region UnityMathematics As Array
```
^ same as above

NoelStephensUnity replied:
^(see above)

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -1386,6 +1653,272 @@ public struct ForFixedStrings
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void WriteValueSafe(Ray2D[] value) => WriteUnmanagedSafe(value);
 
+        #region UnityMathematics As Value Safe
```
^

NoelStephensUnity replied:
^

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferWriter.cs:
```diff
@@ -1386,6 +1653,272 @@ public struct ForFixedStrings
+        public void WriteValueSafe(in uint4x4 value) => WriteUnmanagedSafe(value);
+        #endregion
+
+        #region UnityMathematics As Array Safe
```
^

NoelStephensUnity replied:
^

NoelStephensUnity committed:
update

Finished getting Unity.Mathematics types added to all pertinent serialization pipelines.
Added XML API documentation to all methods within classes/structs/interfaces that had XML API documentation for other types.

NoelStephensUnity committed:
test

Very basic test to validate that all Unity.Mathematics types can be serialized within an INetworkSerializable implementation.

NoelStephensUnity committed:
test

adding some debug information.

NoelStephensUnity committed:
test

Fixing typos in tests for Unity.Mathematics types.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/half-types-and-mathematics-references

NoelStephensUnity committed:
style

removing xml doc param that is no longer valid

NoelStephensUnity committed:
style

whitespace correction

NoelStephensUnity commented:
Most likely going with [PR-2473](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/2473)
_(this PR has too many public API additions)_
